File: backend/src/api/routes/guacamole.py
from fastapi import APIRouter, Depends
from src.services import guac_service
from src.api.auth_deps import get_current_user
import src.models.models as models
router = APIRouter(prefix="/guacamole", tags=["Guacamole"])


# API
@router.get("/token", dependencies=[Depends(get_current_user)], summary="Get Guacamole Token")
def get_guac_token():
    """"Get guacamole token for user"""
    return guac_service.get_guac_token()

@router.get("/connections", dependencies=[Depends(get_current_user)], summary="Get All Connections")
def get_connections():
    return guac_service.get_connections()

@router.get("/connections/{connection_id}", dependencies=[Depends(get_current_user)], summary="Get Connection Details")
def get_connection(connection_id: str):
    return guac_service.get_connection(connection_id)

@router.get("/connections/{connection_id}/url", dependencies=[Depends(get_current_user)], summary="Get Connection URL")
def get_connection_url(connection_id: str):
    return guac_service.get_connection_url(connection_id)

@router.post("/connections/ssh", dependencies=[Depends(get_current_user)], summary="Create SSH connection")
def create_ssh_connection(connection_data: models.CreateSSHConnectionRequest):
    """Create SSH connection"""
    result = guac_service.create_ssh_connection(
        name=connection_data.name,
        hostname=connection_data.hostname,
        username=connection_data.username,
        password=connection_data.password,
        max_connections=connection_data.max_connections,
        max_connections_per_user=connection_data.max_connections_per_user
    )

    return {
        "success": True,
        "message": f"SSH connection '{connection_data.name}' created successfully",
        "connection": result
    }

@router.post("/connections/vnc", dependencies=[Depends(get_current_user)], summary="Create VNC connection")
def create_vnc_connection(connection_data: models.CreateVNCConnectionRequest):
    """Create VNC connection"""
    result = guac_service.create_vnc_connection(
        name=connection_data.name,
        hostname=connection_data.hostname,
        password=connection_data.password,
        port=5900,
        max_connections=connection_data.max_connections,
        max_connections_per_user=connection_data.max_connections_per_user
    )

    return {
        "success": True,
        "message": f"VNC connection '{connection_data.name}' created successfully",
        "connection": result
    }

@router.post("/connections/rdp", dependencies=[Depends(get_current_user)], summary="Create RDP connection")
def create_rdp_connection(connection_data: models.CreateRDPConnectionRequest):
    """Create RDP connection"""
    result = guac_service.create_rdp_connection(
        name=connection_data.name,
        hostname=connection_data.hostname,
        username=connection_data.username,
        password=connection_data.password,
        port=3389,
        max_connections=connection_data.max_connections,
        max_connections_per_user=connection_data.max_connections_per_user
    )

    return {
        "success": True,
        "message": f"VNC connection '{connection_data.name}' created successfully",
        "connection": result
    }

@router.delete("/connections/{connection_id}/delete", dependencies=[Depends(get_current_user)], summary="Delete Connection")
def delete_guacamole_connection(connection_id: str):
    """Delete a Guacamole connection"""
    result = guac_service.delete_connection(connection_id)
    return result

@router.get("/connections/{name}/url", dependencies=[Depends(get_current_user)], summary="Get Connection URL by Name")
def search_connections(name: str):
    """Get connection URL by name"""
    return guac_service.get_connection_url_by_name(name)

File: frontend/src/lib/api/guacamole.ts
import { getAuthToken } from "./auth";

const API_BASE_URL = "http://127.0.0.1:8000";

// Helper function to make authenticated requests
async function authenticatedFetch(url: string, options: RequestInit = {}) {
  const token = getAuthToken();
  if (!token) {
    throw new Error("No authentication token found");
  }

  const response = await fetch(`${API_BASE_URL}${url}`, {
    ...options,
    headers: {
      ...options.headers,
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) {
    const errorData = await response
      .json()
      .catch(() => ({ detail: "Unknown error" }));
    throw new Error(errorData.detail || `HTTP ${response.status}`);
  }

  return response.json();
}

// Guacamole Management
export async function getGuacamoleToken() {
  return authenticatedFetch("/guacamole/token");
}

export async function getConnections() {
  return authenticatedFetch("/guacamole/connections");
}

export async function getConnection(connectionId: string) {
  return authenticatedFetch(`/guacamole/connections/${connectionId}`);
}

export async function getConnectionUrl(connectionId: string) {
  return authenticatedFetch(`/guacamole/connections/${connectionId}/url`);
}

export async function getConnectionUrlByName(name: string) {
  return authenticatedFetch(`/guacamole/connections/${name}/url`);
}

export async function createSSHConnection(
  connectionData: CreateSSHConnectionRequest,
) {
  return authenticatedFetch("/guacamole/connections/ssh", {
    method: "POST",
    body: JSON.stringify(connectionData),
  });
}

export async function createVNCConnection(
  connectionData: CreateVNCConnectionRequest,
) {
  return authenticatedFetch("/guacamole/connections/vnc", {
    method: "POST",
    body: JSON.stringify(connectionData),
  });
}

export async function createRDPConnection(
  connectionData: CreateRDPConnectionRequest,
) {
  return authenticatedFetch("/guacamole/connections/rdp", {
    method: "POST",
    body: JSON.stringify(connectionData),
  });
}

export async function deleteConnection(connectionId: string) {
  return authenticatedFetch(`/guacamole/connections/${connectionId}/delete`, {
    method: "DELETE",
  });
}

// Types
export interface GuacamoleToken {
  token: string;
  expires_at: string;
}

export interface GuacamoleConnection {
  id: string;
  name: string;
  protocol: "ssh" | "vnc" | "rdp";
  hostname: string;
  username?: string;
  max_connections: number;
  max_connections_per_user: number;
  active_connections?: number;
}

export interface CreateSSHConnectionRequest {
  name: string;
  hostname: string;
  username: string;
  password: string;
  max_connections?: number;
  max_connections_per_user?: number;
}

export interface CreateVNCConnectionRequest {
  name: string;
  hostname: string;
  password: string;
  max_connections?: number;
  max_connections_per_user?: number;
}

export interface CreateRDPConnectionRequest {
  name: string;
  hostname: string;
  username: string;
  password: string;
  max_connections?: number;
  max_connections_per_user?: number;
}

export interface ConnectionResponse {
  success: boolean;
  message: string;
  connection: GuacamoleConnection;
}

export interface ConnectionUrlResponse {
  url: string;
  token: string;
}
